File: 215.kthLargestElement/kthLargestElement.cpp
#include <vector>
#include <cstdio>

using namespace std;

class Solution {
   public:
    int _split(vector<int>& nums, int start, int end, int pivot) {
        int lp = start, rp = end;

        while (lp < rp) {
            while (lp < rp && nums[rp] <= pivot) {
                rp--;
            }
            while (lp < rp && nums[lp] >= pivot) {
                lp++;
            }
            if (lp < rp) {
                int tmp = nums[lp];
                nums[lp] = nums[rp];
                nums[rp] = tmp;
            }
        }

        nums[start] = nums[lp];
        nums[lp] = pivot;

        return lp;
    }

    int _quickSelectWrapper(vector<int>& nums, int start, int end, int k) {
        int pivot = nums[start];
        int split = _split(nums, start, end, pivot);
        if (split == k) {
            return nums[k];
        } else if (split < k) {
            return _quickSelectWrapper(nums, split + 1, end, k);
        } else {
            return _quickSelectWrapper(nums, start, split - 1, k);
        }
    }

    int findKthLargest(vector<int>& nums, int k) {
        return _quickSelectWrapper(nums, 0, nums.size() - 1, k - 1);
    }
};

int main() {
    vector<int> test_case = {3, 2, 1, 5, 6, 4};
    int k = 2;
    int res = Solution().findKthLargest(test_case, k);
    printf("%d\n", res);
}

File: 215.kthLargestElement/kthLargestElement.cs
public class Solution {
    private int _quickSelect(int[] nums, int k, int low, int high) {
        int pivot = nums[low];
        int left = low, right = high;
        while (left < right) {
            while (left < right && nums[right] <= pivot) {
                --right;
            }
            while (left < right && nums[left] >= pivot) {
                ++left;
            }
            if (left < right) {
                int tmp = nums[right];
                nums[right] = nums[left];
                nums[left] = tmp;
            }
        }
        nums[low] = nums[left];
        nums[left] = pivot;

        if (k == left) {
            return nums[left];
        } else if (k < left) {
            return _quickSelect(nums, k, low, left - 1);
        } else if (k > right) {
            return _quickSelect(nums, k, left + 1, high);
        }
        return -1; // should never be here
    }
    public int FindKthLargest(int[] nums, int k) {
        return _quickSelect(nums, k - 1, 0, nums.Length - 1);
    }
}
